### PR TITLE
Refactor use async control flow

### PR DESCRIPTION
**Pull request by KSXGitHub**
This PR contains:
- [ ] bugfix
- [ ] feature
- [x] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [ ] yes (*bugfixes and features will not be merged without tests*)
- [x] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:

### Description

Convert a chain of `.then()` into a control flow that uses `await`

### CODE REVIEW

**Comment by shellscape** [comment]
Thanks for the PR :beer:

It's super important to explain the motivation for your change. Please add a description to the pull request to explain why you made the change, what benefits it will have moving forward, and why it's better than the code that previously existed.

**Comment by KSXGitHub** [comment]
> why it's better than the code that previously existed.

* A straightforward control flow is easier to reason about.
* A chain of `.then` increases indent level.

File: src/rollup/index.ts [hunk 1]
```diff
@@ -174,184 +174,170 @@ function assignChunksToBundle(
...
-				function getOutputOptions(rawOutputOptions: GenericConfigObject) {
-					return normalizeOutputOptions(
-						inputOptions as GenericConfigObject,
-						rawOutputOptions,
-						chunks.length > 1,
-						graph.pluginDriver
-					);
-				}
+		)
+		.then(
```
Comment (lukastaegert) [inline src/rollup/index.ts]: Moving to a more async-await based flow is definitely something we want but this is not an improvement now that the different styles are mixed and we have async functions inside Promise chains inside an async function. If you complete the refactoring to use a `try-catch` block to catch the error and get rid of the remaining Promise chains in this section, I could see this merged.

Comment (KSXGitHub) [reply src/rollup/index.ts]: @lukastaegert You want this?

```typescript
	let chunks: Chunk[]

	try {
		await graph.pluginDriver.hookParallel('buildStart', [inputOptions]);
		chunks = await graph.build(
			inputOptions.input as string | string[] | Record<string, string>,
			inputOptions.manualChunks,
			inputOptions.inlineDynamicImports as boolean
		);
	} catch (err) {
		await graph.pluginDriver.hookParallel('buildEnd', [err]);
		throw err;
	}

	await graph.pluginDriver.hookParallel('buildEnd', []);
```

For reference, this is the original:

```typescript
	const chunks = await graph.pluginDriver
		.hookParallel('buildStart', [inputOptions])
		.then(() =>
			graph.build(
				inputOptions.input as string | string[] | Record<string, string>,
				inputOptions.manualChunks,
				inputOptions.inlineDynamicImports as boolean
			)
		)
		.then(
			async chunks => {
				await graph.pluginDriver.hookParallel('buildEnd', []);
				return chunks;
			},
			async err => {
				await graph.pluginDriver.hookParallel('buildEnd', [err]);
				throw err;
			}
		);
```

Comment (lukastaegert) [reply src/rollup/index.ts]: Exactly 👍

Comment (KSXGitHub) [reply src/rollup/index.ts]: Well, I already done that (https://github.com/rollup/rollup/pull/3053/commits/4fcacdfc8fa4d57120278190cecd78aa96cc0bee).

**Comment by lukastaegert** [comment]
As I see you are doing further refactorings on this PR: Is this to be merged or do you expect to do further work on this? Also it would be nice to resolve the conflict with master.

**Comment by KSXGitHub** [comment]
https://github.com/rollup/rollup/pull/3053/commits/4aaf4161a2628ad48c423359a24e1951b3a78075 was my final commit.

> Also it would be nice to resolve the conflict with master.

I must know commits caused this conflicts.

**Pull request merged**